runtime-modules/project-token/src/types.rs

```diff
@@ -76,42 +137,325 @@ pub enum TransferPolicy<Hash> {
+    }
+
+    pub(crate) fn total_amount(&self) -> Balance {
+        self.cliff_amount
+            .saturating_add(self.post_cliff_total_amount)
+    }
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub struct SingleDataObjectUploadParams<JoyBalance> {
+    pub object_creation_params: DataObjectCreationParameters,
+    pub expected_data_size_fee: JoyBalance,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub struct UploadContext<AccountId, BagId> {
+    pub uploader_account: AccountId,
+    pub bag_id: BagId,
+}
+
 #[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
-pub(crate) enum OfferingState {
-    /// Initial idle state
+pub struct WhitelistParams<Hash, SingleDataObjectUploadParams> {
+    /// Whitelist merkle root
+    pub commitment: Hash,
+    /// Optional payload data to upload to storage
+    pub payload: Option<SingleDataObjectUploadParams>,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub struct TokenSaleParams<JoyBalance, Balance, BlockNumber, VestingScheduleParams, AccountId> {
+    /// Account that acts as the source of the tokens on sale
+    pub tokens_source: AccountId,
+    /// Token's unit price in JOY
+    pub unit_price: JoyBalance,
+    /// Number of tokens on sale
+    pub upper_bound_quantity: Balance,
+    /// Optional block in the future when the sale should start (by default: starts immediately)
+    pub starts_at: Option<BlockNumber>,
+    /// Sale duration in blocks
+    pub duration: BlockNumber,
+    /// Optional vesting schedule for all tokens on sale
+    pub vesting_schedule: Option<VestingScheduleParams>,
+    /// Optional total sale purchase amount cap per member
+    pub cap_per_member: Option<Balance>,
+    /// Optional sale metadata
+    pub metadata: Option<Vec<u8>>,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug, Default)]
+pub struct TokenSale<JoyBalance, Balance, BlockNumber, VestingScheduleParams, AccountId> {
+    /// Token's unit price in JOY
+    pub unit_price: JoyBalance,
+    /// Number of tokens still on sale (if any)
+    pub quantity_left: Balance,
+    /// Account that acts as the source of the tokens on sale
+    pub tokens_source: AccountId,
+    /// Block at which the sale started / will start
+    pub start_block: BlockNumber,
+    /// Sale duration (in blocks)
+    pub duration: BlockNumber,
+    /// Optional vesting schedule for all tokens on sale
+    pub vesting_schedule: Option<VestingScheduleParams>,
+    /// Optional total sale purchase amount cap per member
+    pub cap_per_member: Option<Balance>,
+}
+
+impl<JoyBalance, Balance, BlockNumber, AccountId>
+    TokenSale<JoyBalance, Balance, BlockNumber, VestingScheduleParams<BlockNumber>, AccountId>
+where
+    BlockNumber: Saturating + Zero + Copy + Clone + PartialOrd,
+    Balance: Saturating + Clone + Copy + From<u32> + Unsigned + TryInto<u32> + TryInto<u64> + Ord,
+{
+    pub(crate) fn try_from_params<T: Trait>(
+        params: TokenSaleParamsOf<T>,
```

Comment:
Sale with `duration == 0` is just a wasted tx, vesting with `duration == 0` is just a delayed payout (in case `blocks_before_cliff` are not also `== 0`, in which case, it's just an instant payout).

None of these cause any invalid state in the runtime etc., but I can add the checks just to be 100% safe.

Reply:
Introduced `MinSaleDuration` const. In case of vesting, I there is a case for having `blocks_before_cliff == 0` as well as there is a case for having `duration == 0`, so I left it as is.

Reply:
`MinSaleDuration` changed to storage value in https://github.com/Joystream/joystream/pull/3618/commits/ccaf56158ec93ffcee24ecffe81109968858141b